File: 프로그래머스/3/132266. 부대복귀/부대복귀.cpp
#include <string>
#include <vector>
#include <climits>
#include <queue>
using namespace std;

struct Road{
    int next;
    int t;
};

struct xy{
    int now;
    int now_t;
};
struct cmp{
    bool operator()(xy a, xy b){
        return a.now_t > b.now_t;
    }
};

vector<Road> Roads[100001];
int dp[100001];



void dijkstra(int start){
    dp[start] = 0;
    priority_queue<xy,vector<xy>,cmp> pq;
    pq.push({start, 0});
    while(pq.size() !=0){
        xy x = pq.top(); pq.pop();
        for(Road road : Roads[x.now]){
            if(dp[road.next] > dp[x.now] + road.t){
                dp[road.next] = dp[x.now] + road.t;
                pq.push({road.next, dp[road.next]});
            }
        }
    }
}

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
    vector<int> answer;
    
    for(int i = 0; i<roads.size();i++){
        Roads[roads[i][0]].push_back({roads[i][1],1});
        Roads[roads[i][1]].push_back({roads[i][0],1});
    }
    
    for(int i = 0; i<=n; i++)
        dp[i] = INT_MAX;
    
    dijkstra(destination);
    
    for(auto c : sources){
        if(dp[c] == INT_MAX) answer.push_back(-1);
        else answer.push_back(dp[c]);
    }
    
    return answer;
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.io.*;
import java.util.*;
class Solution {
    static ArrayList<Road>[] road;
    static int[] dp;
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        road = new ArrayList[n + 1];
        dp = new int[n+1];
        for(int i = 0; i <= n; i++){
            road[i] = new ArrayList<>();
            dp[i] = Integer.MAX_VALUE;
        }
        
        for(int i = 0; i < roads.length; i++){
            road[ roads[i][0] ].add(new Road( roads[i][1] , 1));
            road[ roads[i][1] ].add(new Road( roads[i][0] , 1));
        }
        
        dijkstra(destination);
        
        int[] answer = new int[ sources.length ];
        int i = 0;
        for(int x : sources){
            answer[i++] = (dp[x] == Integer.MAX_VALUE ? -1 : dp[x]);
        }
        
        return answer;
    }

    static void dijkstra(int i){
        dp[i] = 0;
        PriorityQueue<Road> pq = new PriorityQueue<>((a, b) -> a.dis - b.dis);
        pq.add(new Road(i, 0));
        
        while(!pq.isEmpty()){
            Road x = pq.poll();
            for(Road r : road[x.x]){
                if(dp[r.x] > dp[x.x] + r.dis){
                    dp[r.x] = dp[x.x] + r.dis;
                    pq.add(new Road(r.x, dp[x.x] + r.dis));
                }
            }
        }
    }
    
    static class Road{
        int x, dis;
        Road(int x, int dis){
            this.x = x;
            this.dis = dis;
        }
    }
}